### ISSUE TITLE
Allow failing/skipping Linter bears based on arguments

### ISSUE THREAD

**Issue by encukou** [coala-analyzer/coala#2573]
I am writing a bear that works with `language="Python 2"` or `language="Python 3"`, but should fail (or skip the check?) when asked to process any other language.

I found no good way to do this from `Linter.create_arguments()`: raising an exception is user-unfriendly, and returning anything but an iterable of arguments for the external program is not supported.

Perhaps a special exception or some kind of `validate_arguments()` hook could be added.

**Comment by Makman2** [comment]
I think an exception from `create_arguments` is completely legal, you could also provide `langauge="absolute-total-nonsense-noone-ever-uses`, this would legitimate this even more^^

I think we need to guarantee with our already discussed `FrozenSection` that settings are immutable. This would allow to use the constructor for pre-checks and setting up stuff dependent on the given .coafile .

So I think currently there's no other choice than throwing an exception inside `create_arguments` and coala shows thousands of errors^^

**Comment by sils1297** [comment]
We should definitely catch that exception in linter IMO and error out
nicely.

**Comment by sils1297** [comment]
Oh wait actually there is a choice to do it in check_prerequisites as the section is available there as well right?

**Comment by Makman2** [comment]
if we save the section inside `self.section`, then yes :+1:

**Comment by encukou** [comment]
As a note from a relative outsider, `section` is a very confusing name for a piece of configuration. I think I get where it comes from, but given just the name I'd never guess it holds settings.
Any chance of naming it something like `self.settings`, `self.config`, `self.arguments`?

**Comment by sils1297** [comment]
filed https://github.com/coala-analyzer/coala/issues/2576 for renaming consideration

**Comment by wimpykid26** [comment]
@sils1297 I would like to work on this issue but I do not know how to access the contents of self.section inside Linter.py. Apart from this, would only have to match the file extension that has to be processed against the languages the bear is written for ?

**Comment by mgracer48** [comment]
Ill do this if i get help :)

**Comment by Makman2** [comment]
@wimpykid26 `Linter` is a derivative of bear, so from instance methods you can use `self.section`. And no you don't have to match the file extension, @encukou means a specific setting that's called `language` that shall be checked. We don't have some kind of file-extension-to-language-mapping-functionality yet, this is not that easy^^

But I just remember, you can annotate a bear-parameter with your own type or function, which can throw an error when not conversion is possible. @sils1297 is that correct?

**Comment by sils1297** [comment]
Yep (though I think the error isn't printed in a particular pretty way)

**Comment by Makman2** [comment]
Alright, I think the task for this issue is:
Style up the error printing when the custom conversion function fails (maybe a new custom exception - like `InvalidArgument` or `UnsupportedArgument` - should be introduced for that case),.

**Comment by sils1297** [comment]
we shouldn't rely on a custom exception, it might be a TypeError or
something e.g. when builtins like int or bool are used

**Comment by Makman2** [comment]
ah indeed haven't thought about this :+1: right let's take the default exception thrown when a conversion fails from built-ins :+1:

**Comment by nsgonultas** [comment]
May I work on this one?

**Comment by sils** [comment]
Hi! Welcome to our community and thank you for your interest!

I have sent you an invitation to join our community. This is needed so we can coordinate tasks with you and make sure no two persons are working on the same thing. If you can't find the mail, you should be able to join us at http://github.com/coala-analyzer/!

> Hint: if you're proud to be part of the team, you can make your membership public at https://github.com/orgs/coala/people

We'd love to be able to help you when problems occur. Just join us on [our gitter channel](https://coala.io/chat) and tell us about them :)

**Comment by nsgonultas** [comment]
Invitation accepted. I'll be working on the issue soon.

**Comment by sils** [comment]
It's yours!

**Comment by nsgonultas** [comment]
Can anyone explain what should I do? I'm trying to understand the problem, by reading Linter.py and [Linter Bears section](http://docs.coala.io/en/latest/Developers/Writing_Linter_Bears.html) of the docs but didn't make sense so far :worried:

**Comment by Makman2** [comment]
Honestly I've lost sight a bit. But I believe the major issue is that it's not possible to error out nicely if the arguments provided don't play well together in the eyes of the bear-developer.

It's currently possible to annotate parameters with your own type that can check validity:
```
def myint(value_string):
    i = int(value_string)
    if myint > 1000:
        raise TypeError
    return value_string

def run(self, filename, file, my_argument: myint):
    pass
```

The problem is that you can't check invalid argument *combinations* with this.

I think the best approach is to implement a new `validate_arguments` hook that will be called upon bear instantiation where the user can check if the arguments do work together.

**Comment by sils** [comment]
@Makman2 why? What's wrong with erroring out? In this usecase indeed the simple conversion would suffice.

**Comment by Makman2** [comment]
the user would get a hell list of the same errors / messages that the arguments don't work.

**Comment by nsgonultas** [comment]
So? @Makman2 @sils

**Comment by sils** [comment]
@Makman2 anything can go wrong at any point. IMO erroring out with an exception should be possible anyway. Doesn't have to be param related, I don't understand your problem. It's a builtin concept of the language.

**Comment by nsgonultas** [comment]
What about a function like this:
```python
def validate_arguments(self, expected):
    fails = {}
    for setting, exp_value in expected:
        if self.section.get(setting).value != exp_value:
            fails[setting] = value
    return fails
```
Or it can just return True/False or raise an error? :grin:

**Comment by Makman2** [comment]
@sils but if something's wrong with the passed parameters, it will occur on every file, and the user gets an overflooded terminal^^

**Comment by nsgonultas** [comment]
What about a `validate_settings` method in Bear.py file? Will it work? @Makman2

**Comment by Makman2** [comment]
that's what I suggest :)

**Comment by sils** [comment]
disagree, just throwing an exception is more natural, less complicated and
less automagical

**Comment by Makman2** [comment]
thing is a developer exception means for me something went wrong which should not happen, while providing invalid settings is quite common for a user. Nice little message vs. traceback: I take the message :O
\+ like said again, if some settings are invalid for *all* files the user gets a mass of exceptions, I don't what that from user perspective^^

But what's also possible is to hook the constructor for such kind of validation^^

**Comment by sils** [comment]
> thing is a developer exception means for me: something went wrong which should not happen

We're using this all over the place and raising exceptions to catch them later and make a nice message out of thm.

**Comment by sils** [comment]
I'm not saying to let the exception slip through, linter should catch it.

**Comment by sils** [comment]
But seriously @Makman2 you're making this overly complicated. Exceptions are made for this. Raise, catch, handle.

**Comment by Makman2** [comment]
Imo this should work for all bears, not only linters. If you want an exception for this, we should define one which is just raised on invalid parameters. The problem of getting thousands of messages still persists though^^

**Comment by Nosferatul** [comment]
@nsgonultas are you still working on this?

**Comment by jayvdb** [comment]
Isnt this solved by `Bear.LANGUAGES`.  The coala bear runner can emit a nice message, once, if a bear's `LANGUAGES` does not support the value in the user provided `language=*` parameter .

**Comment by Makman2** [comment]
this is about general settings @jayvdb , for example for linters that only support 2 spaces tabwidth^^ (like some of your issues have required from some bears :D)

**Comment by jayvdb** [comment]
Ah. The opening issue message was about languages.
The problem with most args is that they depend on other args. I.e. google style rules are incompatible with strange tab widths, but even there we could create a custom dynamic style using google+strangetabs. So our new error could actually be NotImplemented.

A good use case would be use use_space, as some langs and linters have mandatory rule for tab vs spaces. I.e. the rule cant be disabled. But, I suspect there isnt any logic possible, as there arnt anti-pattern bears that could be used instead.

So, I think custom meaningful exceptions, with coala catching and cleanly exiting, is an acceptable approach.

**Comment by Makman2** [comment]
I'm still in favour of having a function for that to prevent complete bear execution to avoid thousands of log messages. Especially the nextgen core is not able to stop spawned tasks (okay it should be possible, but makes things certainly harder and I don't want to support this).

**Comment by nsgonultas** [comment]
@Nosferatul no, I'm not. If anybody wants to take it, it's ok for me.

**Comment by sils** [comment]
@nsgonultas in case you didn't already, please do fill coala.io/newform if you have a minute left.
